pkg/controller/cnr/cnr.go

```diff
@@ -0,0 +1,301 @@
+		klog.Infof("Emit Numa Exclusive Anomaly metric...")
+		err = ctrl.emitCNRAnomalyMetric(cnr, reasonNumaExclusiveAnomaly)
+		if err != nil {
+			return err
+		}
+	}
+	// check numa allocatable sum anomaly
+	klog.Infof("Check Numa Allocatable Sum Anomaly...")
+	if ctrl.checkNumaAllocatableSumAnomaly(cnr) {
+		klog.Infof("Emit Numa Allocatable Sum Anomaly metric...")
+		err = ctrl.emitCNRAnomalyMetric(cnr, reasonNumaAllocatableSumAnomaly)
+		if err != nil {
+			return err
+		}
+	}
+
+	// check pod allocation sum anomaly
+	klog.Infof("Check Pod Allocation Sum Anomaly...")
+	if ctrl.checkPodAllocationSumAnomaly(cnr) {
+		klog.Infof("Emit Pod Allocation Sum Anomaly metric...")
+		err = ctrl.emitCNRAnomalyMetric(cnr, reasonPodAllocationSumAnomaly)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// enqueueCNR enqueues the given CNR in the work queue.
+func (ctrl *CNRController) enqueueCNR(cnr *apis.CustomNodeResource) {
+	if cnr == nil {
+		klog.Warning("trying to enqueue a nil cnr")
+		return
+	}
+
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(cnr)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Cound't get key for CNR %+v: %v", cnr, err))
+		return
+	}
+	ctrl.cnrSyncQueue.Add(key)
+}
+
+func (ctrl *CNRController) addCNREventHandler(obj interface{}) {
+	klog.Infof("CNR create event found...")
+	cnr, ok := obj.(*apis.CustomNodeResource)
+	if !ok {
+		klog.Errorf("cannot convert obj to *apis.CNR")
+		return
+	}
+	klog.V(4).Infof("notice addition of cnr %s", cnr.Name)
+
+	ctrl.enqueueCNR(cnr)
+}
+
+func (ctrl *CNRController) updateCNREventHandler(_, newObj interface{}) {
+	klog.Infof("CNR update event found...")
+	cnr, ok := newObj.(*apis.CustomNodeResource)
+	if !ok {
+		klog.Errorf("cannot convert newObj to *apis.CNR")
+		return
+	}
+	klog.V(4).Infof("notice update of cnr %s", cnr.Name)
+
+	// emit cnr pod report lantency metric
+	klog.Infof("Emit CNR Report Lantency metric...")
+	err := ctrl.emitCNRReportLantencyMetric(cnr)
+	if err != nil {
+		klog.Errorf("emit cnr report lantency metric failed: %v", err)
+	}
+
+	ctrl.enqueueCNR(cnr)
+}
+
+func (ctrl *CNRController) updatePodEventHandler(oldObj, newObj interface{}) {
+	klog.Infof("Pod update event found...")
+	oldPod, ok := oldObj.(*corev1.Pod)
+	if !ok {
+		klog.Errorf("cannot convert oldObj to Pod")
+		return
+	}
+	newPod, ok := newObj.(*corev1.Pod)
+	if !ok {
+		klog.Errorf("cannot convert newObj to Pod")
+		return
+	}
+	if oldPod.Spec.NodeName == "" && newPod.Spec.NodeName != "" {
+		klog.Infof("notice pod: %v scheduled to node: %v", newPod.Name, newPod.Spec.NodeName)
+		// record pod scheduled time
+		ctrl.podTimeMap.Store(string(newPod.UID), time.Now())
+	}
+}
+
+// gcPodTimeMap gc podTimeMap which over 1 minutes not used
+func (ctrl *CNRController) gcPodTimeMap() {
+	klog.Infof("gc podTimeMap...")
+	ctrl.podTimeMap.Range(func(key, value interface{}) bool {
+		notUsedTime := time.Now().Sub(value.(time.Time))
+		if notUsedTime > time.Minute {
```

Comment:
if notUsedTime large than the max tolerance report lantency, for example 5min, and use PodIsTerminated func to check whether the pod is terminated,  if not, it can also emit an metricsNameCNRReportLantency metric with value the max tolerance, and add a timeout tag.